### PR TITLE
Canvas provider

### PR DESCRIPTION
**Pull request by wyzlle**
This PR adds required libraries and variables for [canvas](https://www.npmjs.com/package/canvas) projects.

### CODE REVIEW

File: src/nixpacks/plan/generator.rs [hunk 1]
```diff
@@ -207,7 +207,9 @@ impl<'a> NixpacksBuildPlanGenerator<'a> {
                 img => Some(img.to_owned()),
             };
         }
-
+        if NodeProvider::uses_canvas(app) {
+            start_phase.cmd = Some(format!("for pkg in $(ls /nix/store | grep \"util-linux-.*-lib$\"); do export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:\"/nix/store/$pkg/lib\"; done && {}", start_phase.cmd.unwrap_or_default()));
+        }
```
Comment (coffee-cup) [inline src/nixpacks/plan/generator.rs:212]: The generator (or nixpacks in general) shouldn't know anything specific about the providers. This should be moved to the node provider start command.

**Review by coffee-cup — CHANGES_REQUESTED** [review]

File: src/nixpacks/plan/generator.rs (hunk 1 above)
Comment (wyzlle) [reply src/nixpacks/plan/generator.rs:212]: Deployment will crash if someone has `Procfile` in their repo.

Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs:212]: Hmm then need to find a better way to handle Procfile start commands. Or handle util-linux natively in the base image. The node provider should not be imported and used like this. This doesn't seem like a bad thing to have in the base. Do you know how much larger it makes the image?

image

The providers currently have a way of specifying environment variables for the container. We could also expand that functionality so that they can add dynamic things like this?

Comment (wyzlle) [reply src/nixpacks/plan/generator.rs:212]: As mentioned in #183 letting providers read/edit `environment.nix` directly would be better IMO. ([link](https://discourse.nixos.org/t/node2nix-issues/10762/2))

Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs]: The `environment.nix` file doesn't exist when the provider lifecycle methods are run so there is nothing to read/edit yet. I do think though that the providers need a way to inject various Nix snippets into the environment that gets run. Just need to land on an abstraction that will work in the majority of cases.

Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs]: Even just being able to edit `environment.nix` isn't enough. This is an example that would write the library path to a `env.sh` file under the default Nix profile

```nix
{ }:

let pkgs = import <nixpkgs> { };
in with pkgs;
let
  APPEND_LIBRARY_PATH = "${lib.makeLibraryPath [ libGL libuuid ]}";
  myProfile = writeText "env" ''
    export LD_LIBRARY_PATH="${APPEND_LIBRARY_PATH}:$LD_LIBRARY_PATH"
  '';

in buildEnv {
  name = "env";
  paths = [
    (runCommand "env" { } ''
      mkdir -p $out/etc/profile.d
      cp ${myProfile} $out/etc/profile.d/env.sh
    '')

    nodejs
  ];
}
```

But we would still need to somehow load that so those variables are available at runtime.

I think for the `LD_LIBRARY_PATH` variable specifically, we should have a `setup_phase.set_libraries` function that would set everything up automatically.

Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs]: I think this makes adding the libraries available much easier https://github.com/railwayapp/nixpacks/pull/197

File: examples/node-no-lockfile-canvas/index.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+const canvas = require("canvas")
+console.log("Hello from Node");
```
Comment (coffee-cup) [inline examples/node-no-lockfile-canvas/index.js:2]:
```suggestion
console.log("Hello from Node canvas");
```

File: src/providers/node.rs [hunk 3]
```diff
@@ -177,24 +182,34 @@ impl NodeProvider {
     pub fn get_nix_packages(app: &App, env: &Environment) -> Result<Vec<Pkg>> {
         let package_json: PackageJson = app.read_json("package.json")?;
         let node_pkg = NodeProvider::get_nix_node_pkg(&package_json, env)?;
-
+        let mut pkg = vec![node_pkg.clone()];
```
Comment (coffee-cup) [inline src/providers/node.rs:185]:
```suggestion
        let mut pkgs = vec![node_pkg.clone()];
```

File: src/providers/node.rs [hunk 4]
```diff
@@ -177,24 +182,34 @@ impl NodeProvider {
...
         }
+        Ok(pkg)
+    }
+
+    pub fn uses_canvas(app: &App) -> bool {
+        let package_json = app.read_file("package.json").unwrap_or_default();
+        let lock_json = app.read_file("package-lock.json").unwrap_or_default();
+        let yarn_lock = app.read_file("yarn.lock").unwrap_or_default();
+        let pnpm_yaml = app.read_file("pnpm-lock.yaml").unwrap_or_default();
+        package_json.contains("canvas")
```
Comment (coffee-cup) [inline src/providers/node.rs:209]: I feel like this check is a bit too loose? Probably fine but it'll probably end up with quite a few false positives.

**Review by coffee-cup — APPROVED** [review]
Some minor nits but looks good

File: src/providers/node.rs (hunk 4 above)
Comment (wyzlle) [reply src/providers/node.rs]: Improved it a bit. Now, nixpacks will check for `"canvas"` in `package.json` and `/canvas/` in lockfiles.

File: examples/node-no-lockfile-canvas/index.js [hunk 5]
```diff
@@ -0,0 +1,2 @@
+const canvas = require("canvas")
```
Comment (devtomio) [inline examples/node-no-lockfile-canvas/index.js:1]:
```suggestion
require("canvas");
```

Setting it to a variable seems unnecessary as it isn't used.

**Review by devtomio — CHANGES_REQUESTED** [review]

File: examples/node-no-lockfile/index.js [hunk 6]
```diff
@@ -1 +0,0 @@
-console.log("Hello from Node");
```
Comment (Milo123459) [inline examples/node-no-lockfile/index.js:1]: Unrelated change?

File: src/nixpacks/images.rs [hunk 7]
```diff
@@ -1,3 +1,3 @@
-pub static DEFAULT_BASE_IMAGE: &str = "ghcr.io/railwayapp/nixpacks:debian-1654798583";
+pub static DEFAULT_BASE_IMAGE: &str = "ghcr.io/railwayapp/nixpacks:debian-1654800960";
```
Comment (Milo123459) [inline src/nixpacks/images.rs:1]: This also seems unrelated. A rebase would remove this.

File: src/providers/node.rs [hunk 8]
```diff
@@ -36,7 +36,12 @@ impl Provider for NodeProvider {
 
     fn setup(&self, app: &App, env: &Environment) -> Result<Option<SetupPhase>> {
         let packages = NodeProvider::get_nix_packages(app, env)?;
-        Ok(Some(SetupPhase::new(packages)))
+        let mut setup_phase = SetupPhase::new(packages);
+        if NodeProvider::uses_canvas(app) {
```
Comment (Milo123459) [inline src/providers/node.rs:40]: Idea: for the future (I'll implement) a method to check if a `package.json` has a dependency. Should check only `dependencies`.

File: tests/docker_run_tests.rs [hunk 9]
```diff
@@ -238,9 +238,9 @@ fn test_node_custom_version() {
 
 #[test]
 fn test_node_no_lockfile() {
-    let name = simple_build("./examples/node-no-lockfile");
+    let name = simple_build("./examples/node-no-lockfile-canvas");
     let output = run_image(name, None);
-    assert!(output.contains("Hello from Node"));
+    assert!(output.contains("Hello from Node canvas"));
```
Comment (Milo123459) [inline tests/docker_run_tests.rs:243]:
```suggestion
    assert_eq!(output, "Hello from Node canvas");
```

File: examples/node-no-lockfile/index.js (hunk 6 above)
Comment (wyzlle) [reply examples/node-no-lockfile/index.js:1]: Nope.

Comment (Milo123459) [reply examples/node-no-lockfile/index.js:1]: How is this not unrelated?

File: src/nixpacks/images.rs (hunk 7 above)
Comment (wyzlle) [reply src/nixpacks/images.rs:1]: How will a rebase remove this?

Comment (Milo123459) [reply src/nixpacks/images.rs:1]: You're not changing the image in this PR, so a rebase would remove this change

File: examples/node-no-lockfile/index.js (hunk 6 above)
Comment (wyzlle) [reply examples/node-no-lockfile/index.js:1]: We don't need this file anymore.

File: src/nixpacks/images.rs (hunk 7 above)
Comment (wyzlle) [reply src/nixpacks/images.rs:1]: I am updating the image to `debian-1654800960`(canvas requires that). Currently nixpacks uses `debian-1654798583` and that's not the latest release.

Comment (Milo123459) [reply src/nixpacks/images.rs:1]: I thought I saw a PR get merged about updating this?

File: examples/node-no-lockfile/index.js (hunk 6 above)
Comment (Milo123459) [reply examples/node-no-lockfile/index.js:1]: Alright. I have just seen that you have changed the example from no-lockfile to no-lockfile-canvas. I guess that's fine.

File: tests/docker_run_tests.rs (hunk 9 above)
Comment (wyzlle) [reply tests/docker_run_tests.rs:243]: `npm run start` is used as start command here and that logs other stuff too.

File: src/nixpacks/images.rs (hunk 7 above)
Comment (wyzlle) [reply src/nixpacks/images.rs:1]: You saw that right but in that same PR jr made some changes to the base Dockerfile.
